homeworks/12/chat-system/src/Client.java

```diff
@@ -0,0 +1,99 @@
+import java.io.*;
+import java.net.Socket;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+public class Client {
+    private Socket socket;
+    private BufferedReader bufferedReader;
+    private BufferedWriter bufferedWriter;
+    private String nickname;
+
+    public Client(Socket socket, String nickname) {
+        try {
+            this.socket = socket;
+            bufferedReader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+            bufferedWriter = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream()));
+            this.nickname = nickname;
+        } catch (IOException exception){
+            try {
+                bufferedReader.close();
+                bufferedWriter.close();
+                socket.close();
+            }catch (Exception exception1) {
+                System.out.println(exception1);
+            }
+        }
+    }
+
+    public void sendMessage() {
+        try {
+            bufferedWriter.write(nickname);
+            bufferedWriter.newLine();
+            bufferedWriter.flush();
+
+            Scanner scanner = new Scanner(System.in);
+            while (socket.isConnected()) {
+                String message = scanner.nextLine();
+                if(message.startsWith("/msg")) {
+                    String newString = message.substring("/msg".length());
+                    bufferedWriter.write(nickname + ": " + newString);
+                    bufferedWriter.newLine();
+                    bufferedWriter.flush();
+                }
+            }
+        } catch (IOException exception) {
+            try {
+                bufferedReader.close();
+                bufferedWriter.close();
+                socket.close();
+            }catch (Exception exception1) {
+                System.out.println(exception1);
+            }
+        }
+    }
+
+    public void listenForMessage() {
+        new Thread(() -> {
+            String msgFromGroupChat;
+
+            while (socket.isConnected()) {
+                try {
+                    msgFromGroupChat = bufferedReader.readLine();
+                    System.out.println(msgFromGroupChat);
+                } catch (IOException exception) {
+                    try {
+                        bufferedReader.close();
+                        bufferedWriter.close();
+                        socket.close();
+                    }catch (Exception exception1) {
+                        System.out.println(exception1);
+                    }
+                }
+            }
+        }).start();
+    }
+
+    public static void main(String[] args) throws IOException {
+        Scanner scanner = new Scanner(System.in);
+        String nickname = null;
+        String host = null;
+        int port = 0;
+        Socket socket = new Socket();
+        while(host == null || port == 0 || nickname == null) {

```

Comment:
The _while_ loop condition can check if `socket`  is **null**, because you can pass invalid values for `host` and `port` and break the loop without creating the `socket`. For example you can pass _**host = "Gosho"**_, _**port = 70 000**_ and some nickname, let's say _**nickname = "Pesho"**_. "Gosho" is not a valid IP and port 70 000 is not valid as the last port number is 65 535. But you still exit the loop. With invalid arguments in the constructor, `socket` will remain **null**, that's why you can write:

`while(socket == null || nickname == null){`